src/PlusDataCollection/Andor/vtkPlusAndorCamera.cxx

```diff
@@ -0,0 +1,956 @@
+  // GetMostRecentImage() is 32 bit signed variant
+  checkStatus(GetMostRecentImage16(&rawFrame[0], rawFrameSize), "GetMostRecentImage16");
+
+  return PLUS_SUCCESS;
+}
+
+// ----------------------------------------------------------------------------
+void vtkPlusAndorCamera::AddFrameToDataSource(DataSourceArray& ds)
+{
+  for(unsigned i = 0; i < ds.size(); i++)
+  {
+    if(ds[i]->AddItem(&rawFrame[0],
+                      US_IMG_ORIENT_MF,
+                      frameSize, VTK_UNSIGNED_SHORT,
+                      1, US_IMG_BRIGHTNESS, 0,
+                      this->FrameNumber,
+                      currentTime,
+                      UNDEFINED_TIMESTAMP,
+                      nullptr) != PLUS_SUCCESS)
+    {
+      LOG_WARNING("Error adding item to AndorCamera video source " << ds[i]->GetSourceId());
+    }
+    else
+    {
+      LOG_INFO("Success adding item to AndorCamera video source " << ds[i]->GetSourceId());
+    }
+  }
+}
+
+
+// ----------------------------------------------------------------------------
+void vtkPlusAndorCamera::ApplyFrameCorrections()
+{
+  cv::Mat cvIMG(frameSize[0], frameSize[1], CV_16UC1, &rawFrame[0]); // uses rawFrame as buffer
+  cv::Mat floatImage;
+  cvIMG.convertTo(floatImage, CV_32FC1);
+  cv::Mat result;
+
+  cv::subtract(floatImage, cvBiasCorrection, floatImage, cv::noArray(), CV_32FC1);
+  LOG_INFO("Applied constant bias correction");
+
+  // Divide the image by the 32-bit floating point correction image
+  cv::divide(floatImage, cvFlatCorrection, floatImage, 1, CV_32FC1);
+  LOG_INFO("Applied multiplicative flat correction");
+
+  // OpenCV's lens distortion correction
+  cv::undistort(floatImage, result, cvCameraIntrinsics, cvDistanceCoefficients);
+  LOG_INFO("Applied lens distortion correction");
+  result.convertTo(cvIMG, CV_16UC1);
+}
+
+// ----------------------------------------------------------------------------
+PlusStatus vtkPlusAndorCamera::AcquireBLIFrame(int binning, int vsSpeed, int hsSpeed, float exposureTime)
+{
+  WaitForCooldown();
+  AcquireFrame(exposureTime, 0, binning, vsSpeed, hsSpeed);
+  ++this->FrameNumber;
+  AddFrameToDataSource(BLIraw);
+
+  if(this->UseFrameCorrections)
+  {
+    ApplyFrameCorrections();
+    AddFrameToDataSource(BLIrectified);
+  }
+
+  return PLUS_SUCCESS;
+}
+
+// ----------------------------------------------------------------------------
+PlusStatus vtkPlusAndorCamera::AcquireGrayscaleFrame(int binning, int vsSpeed, int hsSpeed, float exposureTime)
+{
+  WaitForCooldown();
+  AcquireFrame(exposureTime, 0, binning, vsSpeed, hsSpeed);
+  ++this->FrameNumber;
+  AddFrameToDataSource(GrayRaw);
+
+  if(this->UseFrameCorrections)
+  {
+    ApplyFrameCorrections();
+    AddFrameToDataSource(GrayRectified);
+  }
+
+  return PLUS_SUCCESS;
+}
+
+// ----------------------------------------------------------------------------
+PlusStatus vtkPlusAndorCamera::AcquireBiasFrame(std::string biasFilePath, int binning, int vsSpeed, int hsSpeed)

```

Comment:
We are going to do a single calibration image which includes both the read bias and dark current. However, this might change in the future so I think we should have the flexibility to take both type of images (Bias- shutter closed, exposure of 0 sec, dark charge- shutter closed, variable exposure) so exposure should be an input to this function. Furthermore, it might be helpful to acquire an image with the shutter open for troubleshooting so the shutter mode should also be an input.

Reply:
It might be appropriate to call this function AcquireCorrectionFrame since it could collect both bias and dark current images.